File: Binary_Search/460.Find K Closest Elements/Solution.java
public class Solution {
    /**
     * @param A: an integer array
     * @param target: An integer
     * @param k: An integer
     * @return: an integer array
     */
    public int[] kClosestNumbers(int[] A, int target, int k) {
        int left = findLowerClosest(A, target);
        int right = left + 1;
        
        int[] results = new int[k];
        for (int i = 0; i < k; i++) {
            if (isLeftCloser(A, target, left, right)) {
                results[i] = A[left];
                left--;
            } else {
                results[i] = A[right];
                right++;
            }
        }
        
        return results;
    }
    
    private boolean isLeftCloser(int[] A, int target, int left, int right) {
        if (left < 0) {
            return false;
        }
        
        if (right >= A.length) {
            return true;
        }
        
        if (target - A[left] != A[right] - target) {
            return target - A[left] < A[right] - target;
        }
        
        return true;
    }
    
    private int findLowerClosest(int[] A, int target) {
        // find the last element smaller than target
        int start = 0, end = A.length - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (A[mid] < target) {
                start = mid;
            } else {
                end = mid;
            }
        }
        
        if (A[end] < target) {
            return end;
        }
        if (A[start] < target) {
            return start;
        }
        
        return -1;
    }
}

File: Binary_Search/460.Find K Closest Elements/Solution.cpp
class Solution
{
  public:
    /**
     * @param A an integer array
     * @param target an integer
     * @param k a non-negative integer
     * @return an integer array
     */
    vector<int> kClosestNumbers(vector<int> &A, int target, int k)
    {
        // Algorithm:
        // 1. Find the first index that A[index] >= target
        // 2. set two pointers left = index - 1 and right = index
        // 3. compare A[left] and A[right] to decide move which pointer

        vector<int> result;
        int index = firstIndex(A, target);
        int left = index - 1, right = index;
        for (int i = 0; i < k; i++)
        {
            // compare A[left] and A[right] to decide move which pointer

            if (left < 0)
            {
                result.push_back(A[right++]);
            }
            else if (right >= A.size())
            {
                result.push_back(A[left--]);
            }
            else
            {
                if (target - A[left] <= A[right] - target)
                {
                    result.push_back(A[left--]);
                }
                else
                {
                    result.push_back(A[right++]);
                }
            }
        }

        return result;
    }

  private:
    int firstIndex(vector<int> &A, int target)
    {
        // use binary search algorithm to find the first index that
        // A[index] >= target

        int start = 0, end = A.size() - 1;
        while (start + 1 < end)
        {
            int mid = start + (end - start) / 2;
            if (A[mid] < target)
            {
                start = mid;
            }
            else
            {
                end = mid;
            }
        }

        if (A[start] >= target)
        {
            return start;
        }

        if (A[end] >= target)
        {
            return end;
        }

        return A.size();
    }
};